packages/dht/src/dht/DhtNode.ts

```diff
@@ -0,0 +1,645 @@
+                PeerID.fromValue(sourcePeer!.peerId),
+                PeerID.fromValue(previousPeer?.peerId || new Uint8Array()))
+            ) {
+                return acc + 1
+            }
+            return acc
+        }, 0)
+    }
+
+    private routeCheck(peerIdToRoute: PeerID, originatorPeerId: PeerID, previousPeerId: PeerID): boolean {
+        return !peerIdToRoute.equals(this.ownPeerId!)
+            && !peerIdToRoute.equals(originatorPeerId)
+            && !peerIdToRoute.equals(previousPeerId)
+    }
+
+    private async getClosestPeersFromContact(contact: DhtPeer): Promise<PeerDescriptor[]> {
+        if (!this.started || this.stopped) {
+            return []
+        }
+        logger.trace(`Getting closest peers from contact: ${contact.peerId.toMapKey()}`)
+        this.outgoingClosestPeersRequestsCounter++
+        this.neighborList!.setContacted(contact.peerId)
+        const returnedContacts = await contact.getClosestPeers(this.ownPeerDescriptor!)
+        this.neighborList!.setActive(contact.peerId)
+        return returnedContacts
+    }
+
+    private onClosestPeersRequestSucceeded(peerId: PeerID, contacts: PeerDescriptor[]) {
+        if (this.ongoingClosestPeersRequests.has(peerId.toMapKey())) {
+            this.ongoingClosestPeersRequests.delete(peerId.toMapKey())
+            const dhtPeers = contacts.map((peer) => {
+                return new DhtPeer(peer, new DhtRpcClient(this.rpcCommunicator!.getRpcClientTransport()))
+            })
+            
+            const oldClosestContact = this.neighborList!.getClosestContactId()
+            
+            dhtPeers.forEach((contact) => this.addNewContact(contact.getPeerDescriptor(), false))
+            
+            if (this.neighborList!.getClosestContactId().equals(oldClosestContact)) {
+                this.noProgressCounter++
+            }
+            else {
+                this.noProgressCounter = 0
+            }
+
+            if (this.ongoingJoinOperation && this.isJoinCompleted()) {
+                this.emit(Event.JOIN_COMPLETED)
+                this.ongoingJoinOperation = false
+            } else {
+                this.findMoreContacts()
+            }
+        }
+    }
+
+    private onClosestPeersRequestFailed(peerId: PeerID, exception: Error) {
+        if (this.ongoingClosestPeersRequests.has(peerId.toMapKey())) {
+            this.ongoingClosestPeersRequests.delete(peerId.toMapKey())
+            logger.debug('onClosestPeersRequestFailed: ' + exception)
+            this.neighborList!.removeContact(peerId)
+            this.findMoreContacts()
+        }
+    }
+
+    isJoinCompleted(): boolean {
+        return (this.neighborList!.getUncontactedContacts(this.PARALLELISM).length < 1
+            || this.noProgressCounter >= 4)
+    }
+
+    joinDht(entryPointDescriptor: PeerDescriptor): Promise<void> {
+        if (!this.started || this.stopped || this.ongoingJoinOperation) {
+            return new Promise((resolve, _reject) => resolve())
```

Comment:
This would work if we annotate `joinDht` with `async` modifier:

```suggestion
            return
```

Also line 435

Reply:
make async with waitForEvent3